### ISSUE TITLE
HtmlGenerator.exe fails with HTTP proxy authentication error (407)

### ISSUE THREAD

**Issue by atifaziz** [KirillOsenkov/SourceBrowser#17]
Suppose you are in the directory `C:\Temp`.

Install `Microsoft.SourceBrowser` ([version 1.0.7](https://www.nuget.org/packages/Microsoft.SourceBrowser/1.0.7) at time of reporting this issue) using NuGet:

    nuget install Microsoft.SourceBrowser -Version 1.0.7

Clone this project on which `HtmlGenerator` will be invoked:

    git clone https://github.com/KirillOsenkov/SourceBrowser

Now run:

    Microsoft.SourceBrowser.1.0.7\tools\HtmlGenerator.exe SourceBrowser\SourceBrowser.sln

You should see the program fail with a HTTP proxy authentication (407) error:

```
13:36:14 Generating website
13:36:14 Reading assembly names from C:\Temp\SourceBrowser\SourceBrowser.sln
13:36:14 Reading assembly names from C:\Temp\SourceBrowser\SourceBrowser.sln complete. Took: 00:00:00.0266311
13:36:14 09/10/2015 13:36:14: First chance exception
System.EntryPointNotFoundException: Unable to find an entry point named 'EventSetInformation' in DLL 'advapi32.dll'.
   at Microsoft.Win32.UnsafeNativeMethods.ManifestEtw.EventSetInformation(Int64 registrationHandle, EVENT_INFO_CLASS informationClass, Void* eventInformation, Int32 informationLength)
   at System.Diagnostics.Tracing.EventProvider.SetInformation(EVENT_INFO_CLASS eventInfoClass, Void* data, Int32 dataSize)
13:36:14 09/10/2015 13:36:14: First chance exception
System.Net.WebException: The remote server returned an error: (407) Proxy Authentication Required.
   at System.Net.HttpWebRequest.GetResponse()
13:36:14 09/10/2015 13:36:14: First chance exception
System.Net.WebException: The remote server returned an error: (407) Proxy Authentication Required.
   at System.Net.HttpWebRequest.GetResponse()
13:36:14 09/10/2015 13:36:14: First chance exception
System.Net.WebException: The remote server returned an error: (407) Proxy Authentication Required.
   at System.Net.WebClient.DownloadDataInternal(Uri address, WebRequest& request)

Unhandled Exception: System.Net.WebException: The remote server returned an error: (407) Proxy Authentication Required.
   at System.Net.WebClient.DownloadDataInternal(Uri address, WebRequest& request)
   at System.Net.WebClient.DownloadString(Uri address)
   at Microsoft.SourceBrowser.HtmlGenerator.Federation..ctor(String[] servers)
   at Microsoft.SourceBrowser.HtmlGenerator.Federation..ctor(IEnumerable`1 servers)
   at Microsoft.SourceBrowser.HtmlGenerator.Federation..ctor()
   at Microsoft.SourceBrowser.HtmlGenerator.Program.IndexSolutions(IEnumerable`1 solutionFilePaths)
   at Microsoft.SourceBrowser.HtmlGenerator.Program.Main(String[] args)
13:36:17 Generating website complete. Took: 00:00:03.2944425
```

**Comment by atifaziz** [comment]
This problem appears to be caused by the design of the [WebProxyAuthenticator](https://github.com/KirillOsenkov/SourceBrowser/blob/6f932364d5c0f3006569a0aa211055e0424c1fcb/src/HtmlGenerator/Utilities/WebProxyAuthenticator.cs) that assumes that if one URL didn't need authentication then all the others wouldn't either. In the [`Federation` where `WebProxyAuthenticator` is used](https://github.com/KirillOsenkov/SourceBrowser/blob/6f932364d5c0f3006569a0aa211055e0424c1fcb/src/HtmlGenerator/Pass1-Generation/Federation.cs#L30), there are two resources downloaded from the URLs http://referencesource.microsoft.com/Assemblies.txt and http://source.roslyn.io/Assemblies.txt, respectively, but the test is done only against the first. On my proxy, the first one did not 407 while the second one did.

**Comment by atifaziz** [comment]
A quick and easy fix is to add an `HtmlGenerator.exe.config` in the same directory as `HtmlGenerator.exe` with the following content:

```xml
<?xml version="1.0"?>
<configuration>
    <system.net>
        <defaultProxy useDefaultCredentials="true"/>
    </system.net>
</configuration>
```

**Comment by KirillOsenkov** [comment]
Thanks, I've just made a fix that:
1. Includes your app.config into the product
2. Calls into authenticator for every URL, not just the first one

Let me know if this fixes it for you. @jrc313 might be interested to code review.

**Comment by KirillOsenkov** [comment]
Also I've just published NuGet 1.0.8 with this fix.

**Comment by jrc313** [comment]
Huh. Who'd have guessed that grabbing text over HTTP would be so fraught with difficulties.

Fix looks good. If you intend to add any additional HTTP requests then I'd be tempted to wrap WebClient.DownloadString in a utility class that will handle authentication or perhaps an extension to WebClient.

**Comment by KirillOsenkov** [comment]
Good point. But this is good for now (YAGNI). Thanks all!

**Comment by atifaziz** [comment]
@KirillOsenkov Thanks for publishing the fix with the `app.config.` It does the job for me but if you intend to keep the `WebProxyAuthenticator` then the [second (code) fix](https://github.com/KirillOsenkov/SourceBrowser/commit/b90abdfc40057d1fb64b7b8c8addf30bfc08ebde) doesn't really solve the problem by trying for each URL. That's because once the first URL passes through, `WebProxyAuthenticator` thinks all's good and won't attempt it for the second one.

Speaking of YAGNI, perhaps it's worth considering dropping `WebProxyAuthenticator` altogether. It's buggy, hard to get right and `app.config` will do the right thing for 98% cases. If someone reports a 407 issue in the future then you could revisit `WebProxyAuthenticator`.

**Comment by jrc313** [comment]
@atifaziz That's a good point. The `WebProxyAuthenticator` stops attempting to authenticate after a successful connection. A simple fix would be to remove the `hasAuthenticated` flag, but I think you're right. If the `app.config` change covers most cases then it makes sense to drop the `WebProxyAuthenticator`.

I can try it from our network on Monday.

**Comment by KirillOsenkov** [comment]
@atifaziz @jrc313 interesting! Thanks for looking into this! Since I don't know anything about proxies, I'll trust you folks to figure out what's the right thing to do here. If app.config works for both of you, we can remove WebProxyAuthenticator. Just let me know what to do or just send a PR.

**Comment by atifaziz** [comment]
I'm afraid the fix in b90abdfc40057d1fb64b7b8c8addf30bfc08ebde doesn't do much since the [config file](https://github.com/KirillOsenkov/SourceBrowser/blob/b90abdfc40057d1fb64b7b8c8addf30bfc08ebde/src/HtmlGenerator/App.config) was never included in the NuGet package.

**Comment by KirillOsenkov** [comment]
Ouch! Forgot to include the file. Thanks for your pull request! I've just accepted and published a new version of NuGet (1.0.9) which includes it.

**Comment by atifaziz** [comment]
Just tested the published 1.0.9 and works like a charm now! Thanks for the quick turnaround on publishing. This issue can be closed from my end.
